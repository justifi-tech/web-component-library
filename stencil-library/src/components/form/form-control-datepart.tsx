import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Element,
  Watch,
} from '@stencil/core';
import IMask, { InputMask } from 'imask';

@Component({
  tag: 'form-control-datepart',
  styleUrl: 'form-control-number.scss',
  shadow: true,
})
export class DatePartInput {
  @Element() el: HTMLElement;

  @Prop() label: string;
  @Prop() name: string;
  @Prop() error: string;
  @Prop() defaultValue: string;
  @Prop() type: 'day' | 'month' | 'year';
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() disabled: boolean;

  private imask: InputMask<any> | null = null;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  componentDidLoad() {
    const inputElement = this.el.shadowRoot.querySelector('input');
    if (inputElement) {
      let maskOptions;

      switch (this.type) {
        case 'day':
          maskOptions = { mask: Number, min: 1, max: 31 };
          break;
        case 'month':
          maskOptions = { mask: Number, min: 1, max: 12 };
          break;
        case 'year':
          maskOptions = {
            mask: Number,
            min: 1900,
            max: new Date().getFullYear(),
          };
          break;
        default:
          throw new Error('Invalid type prop');
      }

      this.imask = IMask(inputElement, maskOptions);

      this.imask.on('accept', () => {
        const rawValue = this.imask.unmaskedValue;
        this.inputHandler(this.name, rawValue);
      });

      inputElement.addEventListener('blur', () => {
        this.formControlBlur.emit();
      });

      this.updateInput(this.defaultValue);
    }
  }

  updateInput(newValue: any) {
    if (this.imask) {
      this.imask.value = newValue;
    }
  }

  disconnectedCallback() {
    this.imask?.destroy();
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          id={this.name}
          name={this.name}
          onBlur={() => this.formControlBlur.emit()}
          part={`input ${this.error && 'input-invalid'}`}
          class={this.error ? 'form-control is-invalid' : 'form-control'}
          type="text"
          disabled={this.disabled}
        />
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
