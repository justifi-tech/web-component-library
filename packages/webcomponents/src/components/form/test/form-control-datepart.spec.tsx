import { newSpecPage } from "@stencil/core/testing";
import { DatePartInput } from "../form-control-datepart";
import * as IMaskPackage from 'imask';

jest.mock('imask', () => ({
  default: jest.fn().mockImplementation(() => ({
    on: jest.fn(),
    destroy: jest.fn(),
    value: '',
    unmaskedValue: '',
  })),
}));

const IMaskSpy = jest.spyOn(IMaskPackage, 'default');

afterEach(() => {
  jest.clearAllMocks();
});

describe('form-control-datepart', () => {
  it('renders correctly with default props', async () => {
    const page = await newSpecPage({
      components: [DatePartInput],
      html: `<form-control-datepart label="Date Part" name="datepart"></form-control-datepart>`,
    });
    expect(page.root).toMatchSnapshot();
  });

  it('should initialize IMask with correct options for "day" type', async () => {
    await newSpecPage({
      components: [DatePartInput],
      html: '<form-control-datepart type="day"></form-control-datepart>',
    });

    expect(IMaskSpy).toHaveBeenCalled();
    expect(IMaskSpy.mock.calls[0][1]).toEqual({ mask: Number, min: 1, max: 31 });
  });

  it('should initialize IMask with correct options for "month" type', async () => {
    // Clear mocks to reset the calls count
    jest.clearAllMocks();

    await newSpecPage({
      components: [DatePartInput],
      html: '<form-control-datepart type="month"></form-control-datepart>',
    });

    expect(IMaskSpy).toHaveBeenCalled();
    expect(IMaskSpy.mock.calls[0][1]).toEqual({ mask: Number, min: 1, max: 12 });
  });

  it('should initialize IMask with correct options for "year" type', async () => {
    // Clear mocks to reset the calls count
    jest.clearAllMocks();

    await newSpecPage({
      components: [DatePartInput],
      html: '<form-control-datepart type="year"></form-control-datepart>',
    });

    expect(IMaskSpy).toHaveBeenCalled();
    expect(IMaskSpy.mock.calls[0][1]).toEqual({ mask: Number, min: 1900, max: new Date().getFullYear() });
  });

  it('emits formControlBlur on input blur', async () => {
    const page = await newSpecPage({
      components: [DatePartInput],
      html: `<form-control-datepart></form-control-datepart>`,
    });

    const blurSpy = jest.fn();
    page.win.addEventListener('formControlBlur', blurSpy);

    const input = page.root.shadowRoot.querySelector('input');
    input.dispatchEvent(new Event('blur', { bubbles: true }));
    await page.waitForChanges();

    expect(blurSpy).toHaveBeenCalled();
  });

  it('calls inputHandler and emits formControlInput on user input', async () => {
    const inputHandlerMock = jest.fn();
    const page = await newSpecPage({
      components: [DatePartInput],
      html: `<form-control-datepart type="day" name="day"></form-control-datepart>`,
    });

    page.rootInstance.inputHandler = inputHandlerMock;
    await page.waitForChanges();

    const inputSpy = jest.fn();
    page.win.addEventListener('formControlInput', inputSpy);

    const input = page.root.shadowRoot.querySelector('input');
    input.value = '15'; // Assuming 'day' type for simplicity
    input.dispatchEvent(new Event('input', { bubbles: true }));
    await page.waitForChanges();

    expect(inputHandlerMock).toHaveBeenCalledWith('day', '15');
    expect(inputSpy).toHaveBeenCalled();
  });

  it('displays error message correctly', async () => {
    const page = await newSpecPage({
      components: [DatePartInput],
      html: `<form-control-datepart error="Error message"></form-control-datepart>`,
    });
    expect(page.root).toMatchSnapshot();
  });
});
