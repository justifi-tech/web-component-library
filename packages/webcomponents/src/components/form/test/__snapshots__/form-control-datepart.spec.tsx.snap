// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-datepart displays error message correctly 1`] = `
<form-control-datepart error="Error message" exportparts="label,input,input-invalid">
  <mock:shadow-root>
    <label class="form-label" part="label"></label>
    <input class="form-control is-invalid" part="input input-invalid" type="text">
    <div class="invalid-feedback">
      Error message
    </div>
  </mock:shadow-root>
</form-control-datepart>
`;

exports[`form-control-datepart renders correctly with default props 1`] = `
<form-control-datepart exportparts="label,input,input-invalid" label="Date Part" name="datepart">
  <mock:shadow-root>
    <label class="form-label" htmlfor="datepart" part="label">
      Date Part
    </label>
    <input class="form-control" id="datepart" name="datepart" part="input undefined" type="text">
  </mock:shadow-root>
</form-control-datepart>
`;
